Release announcement for Kroxylicious 0.4.0 (#16)
kroxylicious/kroxylicious.github.io: _posts/2023-12-13-release-0_4_0.md

---
layout: post
title:  "kroxylicious release 0.4.0"
date:   2023-12-14 00:00:00 +1300
author: "Keith Wall"
author_url: "https://www.github.com/k-wall"
categories:  releases kroxylcious topic-encryption
---

> The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.   The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers.

Reviewer: I'm in two minds about this, its true, but I worry it makes it sound like its finished. Saying its the first iteration isn't very helpful either as we don't actually expect the the filter to change most of the remaining work is around the non functional requirements or reliability. 

Maybe adding another sentence along the lines of 
> We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0](www.github.com/kroxylicious/...) milestone for details

Author: I agree it is important that we get over this isn't finished and folks shouldn't be dropping into production.  I think your words convey that well.  I'll add them.

> The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.   The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers.

Reviewer: From a functional standpoint this is the big news, from a user standpoint I'd say the binary distribution is probably as big news.

The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.

The Kafka Cluster *never* sees the plain text of your records.   You use filter configuration to specify which topic(s) you wish to be encrypted by which key.

> The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  Other Key Management System integrations are planned soon.

Reviewer: ```suggestion
The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  We will provide other selected Key Management System integrations soon, however we are also leveraging our plug-in architecture to allow users to supply their own integrations.
```

The diagram below shows how the system works at a high level.

> ![image](https://github.com/kroxylicious/kroxylicious.github.io/assets/18440250/02b6c1fd-c8a4-495a-80a1-0c7e8f2f97f2)

Reviewer: awesome diagram!

---
layout: post
title:  "kroxylicious release 0.4.0"
date:   2023-12-14 00:00:00 +1300
author: "Keith Wall"
author_url: "https://www.github.com/k-wall"
categories:  releases kroxylcious topic-encryption
---

The Kroxylicious project is very pleased to announce the [release](https://github.com/kroxylicious/kroxylicious/releases/tag/v0.4.0) of Kroxylicious 0.4.0.

> The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.

Reviewer 2: ```suggestion
The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implementation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
```

> The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.

Reviewer 3: Nits: no "the" before Apache Kafka, and "check out" is always two words unless you're referring to the `git` command
```suggestion
The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Check out our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
```

> The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.

Reviewer 3: Nit: capitalise Kafka as it is a noun
```suggestion
The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the Kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
```

> The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.

Reviewer 2: ```suggestion
The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and encrypts the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
```

> The Kafka Cluster *never* sees the plain text of your records.   You use filter configuration to specify which topic(s) you wish to be encrypted by which key.

Reviewer: ```suggestion
The Kafka Cluster *never* sees the plain text of your records. Filter configuration is used to specify which topic(s) should be encrypted by which key.
```

credit to @gracegrimwood for figuring out the right form of words.

Reviewer 3: I have a few more things I would alter here, like using a single space between sentences
```suggestion
The Kafka Cluster *never* sees the plain text of encrypted records. Filter configuration is used to specify which topic(s) are to be encrypted by which key.
```

> The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  We will provide other selected Key Management System integrations soon, however we are also leveraging our plug-in architecture to allow users to supply their own integrations.

Reviewer 3: Nit: rephrasing this so it reads a bit better
```suggestion
The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with a [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm) integration.  We will provide other selected Key Management System integrations soon, however users can leverage the plug-in architecture of Kroxylicious to supply their own integrations.
```

The diagram below shows how the system works at a high level.

> ![image](https://github.com/kroxylicious/kroxylicious.github.io/assets/18440250/09aeb1f4-f420-4d89-b9f7-394473dddc05)

Author: I've updated the diagram to point out Vault can be either on-premise or cloud.

> If the use-case interest you, there are some resources (TODO: link) to help you try out the feature in development.  There will be more information landing on the website soon, but in the meanwhile if you have questions or comments please head over to the [Slack channel](https://kroxylicious.slack.com/).

Reviewer 2: ```suggestion
If the use-case interests you, there are some resources (TODO: link) to help you try out the feature in development.  There will be more information landing on the website soon, but in the meanwhile if you have questions or comments please head over to the [Slack channel](https://kroxylicious.slack.com/).
```

> If this use-case interests you, a [short demo](https://asciinema.org/a/627519) that shows the feature in action will help you understand more and there's some [interim instructions[(https://github.com/kroxylicious/kroxylicious/blob/f5f62a48d1db67309eee1696bb7a70dbb83ba622/kubernetes-examples/envelope-encryption/README.md) to help you try out the feature in development.  There will be more information landing on the website soon, but meanwhile, if you have questions or comments please head over to the [Slack channel](https://kroxylicious.slack.com/).

Author: My intent is that https://github.com/kroxylicious/kroxylicious/pull/825 is merged then this link it updated to refer to main.    For the next release https://github.com/kroxylicious/kroxylicious/issues/827 should be resolved, so we'd be able to reference the example on tag (hopefully it'll be superseded by quickstart content by then too).

categories:  releases kroxylcious topic-encryption
---

The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.

> The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.

Author: I used EnvelopeEncryption here as that's what we've called the filter - so the argument being consistency.  You could argue that the fact it is using envelope encryption is an implementation detail.  The important thing is what it does - TopicEncryption, or EncryptionAtRest.

Reviewer: I think you've expressed it well. We are delivering encryption-at-rest via an envelope encryption filter.
